lld/MachO/ObjC.cpp

```diff
@@ -1183,26 +1193,66 @@ void ObjcCategoryMerger::eraseMergedCategories() {
   // the references to the ones we merged.
   generateCatListForNonErasedCategories(catListToErasedOffsets);
 
+  // We use erasedIsecs below to track erased sections so we can later remove
+  // references to it.
+  std::unordered_set<InputSection *> erasedIsecs;
+  erasedIsecs.reserve(categoryMap.size());
+
   // Erase the old method lists & names of the categories that were merged
   for (auto &mapEntry : categoryMap) {
     for (InfoInputCategory &catInfo : mapEntry.second) {
       if (!catInfo.wasMerged)
         continue;
 
+      erasedIsecs.insert(catInfo.catBodyIsec);
+      erasedIsecs.insert(catInfo.catListIsec);
```

Comment:
I'm trying to follow if `erasedIsecs` is inserted when `eraseISec` is called. So, is this line expected?

Reply:
I am not sure I understand the question. 
For a category we need to erase the body, the catlist reference and all lists. 
Here, the body and list isec is marked as to be erased. They are not marked anywhere else. So both have to be here. 

`eraseISec` function does not insert into `erasedIsecs` - it does not know it exists.

Reply:
The confusion (between `eraseISec` and `eraseISecs`) is exactly what I'm asking. I saw you call `eraseISec` below once for the body while inserting both body and list.  However, I'm also seeing you do these operations in sync in `tryEraseDefinedAtIsecOffset`. 
```
    eraseISec(cisec);
    erased.insert(cisec);
```

This looks confusing and error-prone. I wonder if we can have a single call to handle these together somehow.

Reply:
Move the insert into `eraseISec()` call which simplified the change a lot.